design/velero-upgrade.md

```diff
@@ -0,0 +1,80 @@
+- Support future evolvements. For example, in future, Velero will have various data movement features, for which we will add new moudles inevitably. Then we can add enhancement incrementally to ```velero upgrade```, because inside the command, we can do whatever complex things 
+
+The ```velero upgrade``` command will fully replace the existing manual steps in the aforementioned Velero document.
+
+## Design
+
+At present, ```velero upgrade``` will do below things:
+- Get the new version from the client binary itself, that is, the upgrade target version
+- Get the old version from the existing velero server, that is, the current version
+- Retrieve Velero server deployment object
+- Retrieve Velero daemonset object
+- Iterate from current version through target version, apply the changes (i.e., the image, name and parameter changes) to the deployment object once per version. For example, when upgrading from 1.9 to 1.11, we need to apply the changes between 1.9 and 1.10 first, then apply the changes between 1.10 and 1.11. See the details below for how to drive the changes apply
+- Do the same changes apply to Velero daemonset object as the deployment object
+- Update the Velero server deployment from the modified deployment object
+- Delete the existing Velero daemonset
+- Create the new Velero daemonset from the modified daemonset object
+- Extract the new CRDs from the client binary and apply them
+
+### Action Table
+We create below action functions to apply changes for each target version, the number in the function name indicates the target version:  
+```
+func UpgradeV10(deploy *v1.Deployment, ds *v1.DaemonSet) error
+func UpgradeV11(deploy *v1.Deployment, ds *v1.DaemonSet) error
+func UpgradeV12(deploy *v1.Deployment, ds *v1.DaemonSet) error
+```
+
+Then we put these functions to a vector, so that the action functions could be retrieved by indexes:  
+```
+type UpgradeAction struct {
+	Target string
+	Action func(*v1.Deployment, *v1.DaemonSet) error
+}
+
+var UpgradeActions = []UpgradeAction {
+	{
+		"v1.10",
+		UpgradeV10,
+	},
+	{
+		"v1.11",
+		UpgradeV11,
+	},
+	{
+		"v1.12",
+		UpgradeV12,
+	},	
+}
+```
+So ```UpgradeActions[0]``` means the action to upgrade from older versions to v1.10; ```UpgradeActions[1]``` means the action to upgrade from v1.10 to v1.11.  
+Then we match the current version and target version in ```UpgradeActions``` to get a begin-end index pair.  
+Finally, we call the action functions between the begin index and end index.  
+
+As we can see above, all the required information is self-retrieved, and the upgrade actions are self-driven, so we don't need to add any extra parameters to ```velero upgrade```.  
+
+## Non-Goals
+As the best practise, there should be some kind of recovery measure so that when the upgrade fails, the original environment could be recovered. This design doesn't include it, we could handle this in the future enhancements.  
```

Comment:
So if `velero upgrade` fails we will ask user to uninstall velero and install the latest version, right?

Reply:
Yes, this situation is the same as the existing manual upgrade approach, once it fails, users need to either do the uninstall-install steps or try to remedy it based on the failed installation.

Reply:
Added some explanation for the failed case.